### ISSUE TITLE
Allow rolling back to a cancelled deployment

### ISSUE THREAD

**Issue by kargakis** [openshift/origin#9158]
Scenario:

I start a new deployment, edit it before it is finished, deploy the new one.
Later, I try to rollback to the first but I cannot because it was never tagged as Complete.
We should allow rolling back to cancelled deployments.

**Comment by matthyx** [comment]
@kargakis can I look at this one too? Or should you discuss with others to check its validity first?

**Comment by kargakis** [comment]
@mfojtik

**Comment by matthyx** [comment]
Oops, sorry, I looked at who opened it, rather than the owner...
@mfojtik can I look at this issue? Is it still valid?

**Comment by mfojtik** [comment]
does this problem exists only if you cancel the initial (rev=1) deployment? With cancel I mean you edited it during it being deployed, which cancelled it, and you ended up with (rev=2) which you are not happy with and you would want to roll it back to rev=1 ?

@kargakis do we support rollback, or retry a specific version of deployment? I mean you probably want to rollback to something you know it succeded (and should succeed again), instead of rolling back to something that was cancelled.

**Comment by kargakis** [comment]
We don't know if something that got cancelled succeeded or not because it
was canncelled.

**Comment by mfojtik** [comment]
@kargakis  don't we have "rollout retry" for that? if I cancel something, it sounds good to me to retry it... but I don't think I want to rollback to cancelled revision because I don't know whether it succeed or fail.

**Comment by matthyx** [comment]
@mfojtik I don't see the retry subcommand for rollout...
```
$ oc help rollout
Manage deployments.

Usage:
  oc rollout SUBCOMMAND [options]

Available Commands:
  history     view rollout history
  pause       Mark the provided resource as paused
  resume      Resume a paused resource
  undo        undoes a previous rollout

Use "oc <command> --help" for more information about a given command.
Use "oc options" for a list of global command-line options (applies to all commands).
```

**Comment by matthyx** [comment]
@kargakis @mfojtik @pweil- do we still need this issue then? I think we could allow this rollback and then let the health probes ensure we don't break deployments...

**Comment by mfojtik** [comment]
@matthyx sounds good to me

**Comment by matthyx** [comment]
@mfojtik after doing my initial tests, turns out this issue has already been fixed (at least in origin 1.4):
I create a project and deploy an application inside it
```
[mbertschy@devops origin]$ oc new-project ma4
Now using project "ma4" on server "https://manage.playground.sicpa-ch7.dev:443".

You can add applications to this project with the 'new-app' command. For example, try:

    oc new-app centos/ruby-22-centos7~https://github.com/openshift/ruby-ex.git

to build a new example application in Ruby.
[mbertschy@devops origin]$ oc rollout history dc/mvn-avatar
deploymentconfigs "mvn-avatar"
REVISION        STATUS          CAUSE
1               Complete        config change
```
I edit the deployment to add a new environment variable version=2
```
[mbertschy@devops origin]$ oc edit dc mvn-avatar
deploymentconfig "mvn-avatar" edited
[mbertschy@devops origin]$ oc rollout history dc/mvn-avatar
deploymentconfigs "mvn-avatar"
REVISION        STATUS          CAUSE
1               Complete        config change
2               Running         config change
```
I quickly change it to version=3
```
[mbertschy@devops origin]$ oc edit dc mvn-avatar
deploymentconfig "mvn-avatar" edited
[mbertschy@devops origin]$ oc rollout history dc/mvn-avatar
deploymentconfigs "mvn-avatar"
REVISION        STATUS          CAUSE
1               Complete        config change
2               Failed          newer deployment was found running
3               Complete        config change

[mbertschy@devops origin]$ oc rsh mvn-avatar-3-jp5xg env | grep version
version=3
```
Please note that version 2 is marked as Failed. I rollback to it anyway
```
[mbertschy@devops origin]$ oc rollback mvn-avatar-2
#4 rolled back to mvn-avatar-2
[mbertschy@devops origin]$ oc rollout history dc/mvn-avatar
deploymentconfigs "mvn-avatar"
REVISION        STATUS          CAUSE
1               Complete        config change
2               Failed          newer deployment was found running
3               Complete        config change
4               Complete        config change

[mbertschy@devops origin]$ oc rsh mvn-avatar-4-4bgw0 env | grep version
version=2
```
As you can see, we're successfully back to version 2.

I think you can close this issue :-)
